xc/device.ml

```diff
@@ -1444,22 +1461,34 @@ module Vusb = struct
       end
     in
     let is_running = Qemu.is_running ~xs domid in
-    match is_running with
-    | true  ->
-      let cmd = Qmp.(Device_add
-                       ( "usb-host"
-                       , id
-                       , Some (get_bus version, hostbus, hostport)
-                       ))
-      in qmp_send_cmd domid cmd |> ignore
-    | _ -> ()
-
-  let vusb_unplug ~xs ~domid ~id =
+    if is_running then
+      begin
+        (* Need to reset USB device before passthrough to vm according to CP-24616.
+           Also need to do deprivileged work in usb_reset script if QEMU is deprivileged.
+        *)
+        begin match Qemu.pid ~xs domid with
+          | Some pid -> call_usb_reset hostbus hostport domid pid "attach" priv
+          | _ -> failwith (Printf.sprintf "qemu pid does not exist for vm %s" (string_of_int domid))
+        end;
+        let cmd = Qmp.(Device_add
+                         ( "usb-host"
+                         , id
+                         , Some (get_bus version, hostbus, hostport)
+                         ))
+        in qmp_send_cmd domid cmd |> ignore
+      end
+
+  let vusb_unplug ~xs ~priv ~domid ~id ~hostbus ~hostport=
     debug "Vusb unplugged: vusb device unplugged";
     let is_running = Qemu.is_running ~xs domid in
-    match is_running with
-    | true  -> qmp_send_cmd domid Qmp.(Device_del id) |> ignore
-    | _     -> ()
+    if is_running then
+      begin
+        qmp_send_cmd domid Qmp.(Device_del id) |> ignore;
+        (*Call vusb cleanup in usb_reset.py*)
+        match Qemu.pid ~xs domid with
+        | Some pid -> call_usb_reset hostbus hostport domid pid "detach" priv
+        | _ -> failwith (Printf.sprintf "qemu pid does not exist for vm %s" (string_of_int domid))
```

Comment:
Here as well. I think we should use tractable exceptions in place of `failwith`s. It will be easier to deal with them at the xapi report level